README.md
proxy
=====
### An HTTP proxy written with Node.js (think Squid)
[![Build Status](https://travis-ci.org/TooTallNate/proxy.svg?branch=master)](https://travis-ci.org/TooTallNate/proxy)

This module provides standard "HTTP proxy" logic. You can script your own server
using the `proxy` server API. Be sure to take a look at the "Examples" section
below.

There is also a companion `proxy(1)` CLI tool, which spawns an HTTP(s) proxy
server with the specified options.

You could think of `proxy(1)` as similar to some of the other popular open
source HTTP proxy software:

 * [Squid][]
 * [Privoxy][]
 * [Apache][] with [`mod_proxy`][mod_proxy]
 * [More…](http://wikipedia.org/wiki/Proxy_server#Web_proxy_servers)


Installation
------------

Install with `npm`:

``` bash
$ npm install proxy
```

If you would like to have the `proxy(1)` CLI program in your `$PATH`, then
install "globally":

``` bash
$ npm install -g proxy
```


Examples
--------

#### Basic HTTP(s) proxy server

A basic HTTP(s) server with all the default options. All requests are allowed.
CONNECT HTTP method works as well.

``` js
var http = require('http');
var setup = require('proxy');

var server = setup(http.createServer());
server.listen(3128, function () {
  var port = server.address().port;
  console.log('HTTP(s) proxy server listening on port %d', port);
});
```


CLI Tool Examples
-----------------

The `proxy(1)` CLI tool can be used to spawn HTTP(s) proxy server instances with
various options.

#### Port to bind to

Pass the `-p`/`--port` option to with a port number to specify a TCP port to
bind to. Defaults to __3128__ if none is specified.

``` bash
$ proxy --port 8080
```

#### Custom `Proxy-Authenticate` command

Pass the `-a`/`--authenticate` switch with a command to execute when the client
`Proxy-Authorization` header is given. This command determines whether or not the
request is authorized based on the "exit code" of the command.

The relevant request authentication information is passed in as
`PROXY_AUTH_USERNAME`, `PROXY_AUTH_PASSWORD` and `PROXY_AUTH_SCHEME` environment
variables.

For example, to authorize "Basic" authentication with username "foo" and
password "bar":

``` bash
$ proxy --authenticate 'if \
    [ "$PROXY_AUTH_USERNAME" = "foo" ] && \
    [ "$PROXY_AUTH_PASSWORD" = "bar" ]; \
      then exit 0; \
    fi; \
    exit 1;'
```


License
-------

(The MIT License)

Copyright (c) 2013 Nathan Rajlich &lt;nathan@tootallnate.net&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


[Squid]: http://www.squid-cache.org/
[Privoxy]: http://www.privoxy.org/
[Apache]: http://www.apache.org/
[mod_proxy]: http://httpd.apache.org/docs/current/mod/mod_proxy.html


History.md

0.2.4 / 2016-01-18
==================

  * travis: test node v0.8, v0.10, and v0.12, v1, v2, v3, v4, and v5
  * README: use SVG for Travis-CI badge
  * package: update all the deps

0.2.3 / 2014-04-04
==================

  * package: update "mocha", remove from dependencies

0.2.2 / 2014-04-03
==================

  * History: match `git changelog` styling
  * package: update outdated deps
  * proxy: remove double semicolon
  * proxy: refactor to make the named functions be top-level
  * proxy: add a few debug() calls

0.2.1 / 2013-09-16
==================

  * Close the socket after a CONNECT authorization request

0.2.0 / 2013-09-10
==================

  * Fix calling the setup() function without any arguments
  * Initial integration of "commander" for the `proxy(1)` program

0.1.0 / 2013-09-09
==================

  * Added "Via" header
  * Added "X-Forwared-For" header
  * Strip hop-by-hop headers from requests and responses
  * Better tests (no longer reach the internet)
  * Add `.travis.yml` file
  * Proxy headers as original casing with node >= v0.11.6 (`rawHeaders`)
  * bin: set the `process.title` to "proxy"

0.0.1 / 2013-09-07
==================

  * Initial release


test/test.js

/**
 * Module dependencies.
 */

var fs = require('fs');
var net = require('net');
var path = require('path');
var http = require('http');
var https = require('https');
var assert = require('assert');
var setup = require('../');
var stream = require('stream');

describe('proxy', function () {

  var proxy;
  var proxyPort;

  var transformProxy;
  var transformProxyPort;
  var responseStream = new stream.PassThrough();
  var requestStream = new stream.PassThrough();

  var server;
  var serverPort;

  before(function (done) {
    // plain old proxy.
    function createProxy(cb) {
      proxy = setup(http.createServer());
      proxy.listen(function () {
        proxyPort = proxy.address().port;
        return cb();
      });
    }

    // proxy that transforms HTTP stream.
    function createTransformProxy(cb) {
      transformProxy = setup(http.createServer(), {
        transformResponse: responseStream,
        transformRequest: requestStream
      });
      transformProxy.listen(function () {
        transformProxyPort = transformProxy.address().port;
        return cb();
      });
    }

    createProxy(function() {
      createTransformProxy(function() {
        return done();
      });
    });
  });

  before(function (done) {
    // setup target server
    server = http.createServer();
    server.listen(function () {
      serverPort = server.address().port;
      done();
    });
  });

  after(function (done) {
    proxy.once('close', function () { done(); });
    proxy.close();
  });

  after(function (done) {
    transformProxy.once('close', function () { done(); });
    transformProxy.close();
  });

  after(function (done) {
    server.once('close', function () { done(); });
    server.close();
  });

  it('should proxy HTTP GET requests', function (done) {
    var gotData = false;
    var gotRequest = false;
    var host = '127.0.0.1:' + serverPort;
    server.once('request', function (req, res) {
      gotRequest = true;
      // ensure headers are being proxied
      assert(req.headers['user-agent'] == 'curl/7.30.0');
      assert(req.headers.host == host);
      assert(req.headers.accept == '*/*');
      res.end();
    });

    var socket = net.connect({ port: proxyPort });
    socket.once('close', function () {
      assert(gotData);
      assert(gotRequest);
      done();
    });
    socket.once('connect', function () {
      socket.write(
        'GET http://' + host + '/ HTTP/1.1\r\n' +
        'User-Agent: curl/7.30.0\r\n' +
        'Host: ' + host + '\r\n' +
        'Accept: */*\r\n' +
        'Proxy-Connection: Keep-Alive\r\n' +
        '\r\n');
    });
    socket.setEncoding('utf8');
    socket.once('data', function (data) {
      assert(0 == data.indexOf('HTTP/1.1 200 OK\r\n'));
      gotData = true;
      socket.destroy();
    });
  });

  it('should allow the request and response streams to be transformed', function (done) {
    var host = '127.0.0.1:' + serverPort;
    server.once('request', function (req, res) {
      res.end();
    });

    var socket = net.connect({ port: transformProxyPort });
    socket.once('close', function () {
      assert(requestStream._writableState.ended);
      assert(responseStream._writableState.ended);
      done();
    });
    socket.once('connect', function () {
      socket.write(
        'GET http://' + host + '/ HTTP/1.1\r\n' +
        'User-Agent: curl/7.30.0\r\n' +
        'Host: ' + host + '\r\n' +
        'Accept: */*\r\n' +
        'Proxy-Connection: Keep-Alive\r\n' +
        '\r\n');
    });
    socket.setEncoding('utf8');
    socket.once('data', function (data) {
      socket.destroy();
    });
  });

  it('should establish connection for CONNECT requests', function (done) {
    var gotData = false;
    var socket = net.connect({ port: proxyPort });
    socket.once('close', function () {
      assert(gotData);
      done();
    });
    socket.once('connect', function () {
      var host = '127.0.0.1:' + serverPort;
      socket.write(
        'CONNECT ' + host + ' HTTP/1.1\r\n' +
        'Host: ' + host + '\r\n' +
        'User-Agent: curl/7.30.0\r\n' +
        'Proxy-Connection: Keep-Alive\r\n' +
        '\r\n');
    });
    socket.setEncoding('utf8');
    socket.once('data', function (data) {
      assert(0 == data.indexOf('HTTP/1.1 200 Connection established\r\n'));
      gotData = true;
      socket.destroy();
    });
  });

  describe('authentication', function () {
    function clearAuth () {
      delete proxy.authenticate;
    }

    before(clearAuth);
    after(clearAuth);

    it('should invoke the `server.authenticate()` function when set', function (done) {
      var auth = 'Basic Zm9vOmJhcg==';
      var called = false;
      proxy.authenticate = function (req, fn) {
        assert(auth == req.headers['proxy-authorization']);
        socket.destroy();
        called = true;
      };
      var socket = net.connect({ port: proxyPort });
      socket.once('close', function () {
        assert(called);
        done();
      });
      socket.once('connect', function () {
        socket.write(
          'GET / HTTP/1.1\r\n' +
          'Proxy-Authorization: ' + auth + '\r\n' +
          '\r\n');
      });
    });

    it('should provide the HTTP client with a 407 response status code', function (done) {
      proxy.authenticate = function (req, fn) {
        // reject everything
        fn(null, false);
      };
      var gotData = false;
      var socket = net.connect({ port: proxyPort });
      socket.once('close', function () {
        assert(gotData);
        done();
      });
      socket.once('connect', function () {
        socket.write('GET / HTTP/1.1\r\n\r\n');
      });
      socket.setEncoding('utf8');
      socket.once('data', function (data) {
        assert(0 == data.indexOf('HTTP/1.1 407'));
        gotData = true;
        socket.destroy();
      });
    });

    it('should close the socket after a CONNECT request\'s 407 response status code', function (done) {
      proxy.authenticate = function (req, fn) {
        // reject everything
        fn(null, false);
      };
      var gotData = false;
      var socket = net.connect({ port: proxyPort });
      socket.once('close', function () {
        assert(gotData);
        done();
      });
      socket.once('connect', function () {
        socket.write('CONNECT 127.0.0.1:80 HTTP/1.1\r\n\r\n');
      });
      socket.setEncoding('utf8');
      socket.once('data', function (data) {
        assert(0 == data.indexOf('HTTP/1.1 407'));
        gotData = true;
      });
    });

  });

});


proxy.js

/**
 * Module dependencies.
 */

var net = require('net');
var url = require('url');
var http = require('http');
var assert = require('assert');
var debug = require('debug')('proxy');

// log levels
debug.request = require('debug')('proxy ← ← ←');
debug.response = require('debug')('proxy → → →');
debug.proxyRequest = require('debug')('proxy ↑ ↑ ↑');
debug.proxyResponse = require('debug')('proxy ↓ ↓ ↓');

// hostname
var hostname = require('os').hostname();

// proxy server version
var version  = require('./package.json').version;

/**
 * Module exports.
 */

module.exports = setup;
var sOptions = {};

/**
 * Sets up an `http.Server` or `https.Server` instance with the necessary
 * "request" and "connect" event listeners in order to make the server act as an
 * HTTP proxy.
 *
 * @param {http.Server|https.Server} server
 * @param {Object} options
 * @api public
 */

function setup (server, options) {
  if (!server) server = http.createServer();
  server.on('request', onrequest);
  server.on('connect', onconnect);
  sOptions = options
  return server;
}

/**
 * 13.5.1 End-to-end and Hop-by-hop Headers
 *
 * Hop-by-hop headers must be removed by the proxy before passing it on to the
 * next endpoint. Per-request basis hop-by-hop headers MUST be listed in a
 * Connection header, (section 14.10) to be introduced into HTTP/1.1 (or later).
 */

var hopByHopHeaders = [
  'Connection',
  'Keep-Alive',
  'Proxy-Authenticate',
  'Proxy-Authorization',
  'TE',
  'Trailers',
  'Transfer-Encoding',
  'Upgrade'
];

// create a case-insensitive RegExp to match "hop by hop" headers
var isHopByHop = new RegExp('^(' + hopByHopHeaders.join('|') + ')$', 'i');

/**
 * Iterator function for the request/response's "headers".
 * Invokes `fn` for "each" header entry in the request.
 *
 * @api private
 */

function eachHeader (obj, fn) {
  if (Array.isArray(obj.rawHeaders)) {
    // ideal scenario... >= node v0.11.x
    // every even entry is a "key", every odd entry is a "value"
    var key = null;
    obj.rawHeaders.forEach(function (v) {
      if (key === null) {
        key = v;
      } else {
        fn(key, v);
        key = null;
      }
    });
  } else {
    // otherwise we can *only* proxy the header names as lowercase'd
    var headers = obj.headers;
    if (!headers) return;
    Object.keys(headers).forEach(function (key) {
      var value = headers[key];
      if (Array.isArray(value)) {
        // set-cookie
        value.forEach(function (val) {
          fn(key, val);
        });
      } else {
        fn(key, value);
      }
    });
  }
}

/**
 * HTTP GET/POST/DELETE/PUT, etc. proxy requests.
 */

function onrequest (req, res) {
  debug.request('%s %s HTTP/%s ', req.method, req.url, req.httpVersion);
  var server = this;
  var socket = req.socket;

  // pause the socket during authentication so no data is lost
  socket.pause();

  authenticate(server, req, function (err, auth) {
    socket.resume();
    if (err) {
      // an error occured during login!
      res.writeHead(500);
      res.end((err.stack || err.message || err) + '\n');
      return;
    }
    if (!auth) return requestAuthorization(req, res);
    var parsed = url.parse(req.url);

    // proxy the request HTTP method
    parsed.method = req.method;

    // setup outbound proxy request HTTP headers
    var headers = {};
    var hasXForwardedFor = false;
    var hasVia = false;
    var via = '1.1 ' + hostname + ' (proxy/' + version + ')';

    parsed.headers = headers;
    eachHeader(req, function (key, value) {
      debug.request('Request Header: "%s: %s"', key, value);
      var keyLower = key.toLowerCase();

      if (!hasXForwardedFor && 'x-forwarded-for' === keyLower) {
        // append to existing "X-Forwarded-For" header
        // http://en.wikipedia.org/wiki/X-Forwarded-For
        hasXForwardedFor = true;
        value += ', ' + socket.remoteAddress;
        debug.proxyRequest('appending to existing "%s" header: "%s"', key, value);
      }

      if (!hasVia && 'via' === keyLower) {
        // append to existing "Via" header
        hasVia = true;
        value += ', ' + via;
        debug.proxyRequest('appending to existing "%s" header: "%s"', key, value);
      }

      if (isHopByHop.test(key)) {
        debug.proxyRequest('ignoring hop-by-hop header "%s"', key);
      } else {
        var v = headers[key];
        if (Array.isArray(v)) {
          v.push(value);
        } else if (null != v) {
          headers[key] = [ v, value ];
        } else {
          headers[key] = value;
        }
      }
    });

    // add "X-Forwarded-For" header if it's still not here by now
    // http://en.wikipedia.org/wiki/X-Forwarded-For
    if (!hasXForwardedFor) {
      headers['X-Forwarded-For'] = socket.remoteAddress;
      debug.proxyRequest('adding new "X-Forwarded-For" header: "%s"', headers['X-Forwarded-For']);
    }

    // add "Via" header if still not set by now
    if (!hasVia) {
      headers.Via = via;
      debug.proxyRequest('adding new "Via" header: "%s"', headers.Via);
    }

    // custom `http.Agent` support, set `server.agent`
    var agent = server.agent;
    if (null != agent) {
      debug.proxyRequest('setting custom `http.Agent` option for proxy request: %s', agent);
      parsed.agent = agent;
      agent = null;
    }

    if (null == parsed.port) {
      // default the port number if not specified, for >= node v0.11.6...
      // https://github.com/joyent/node/issues/6199
      parsed.port = 80;
    }

    if ('http:' != parsed.protocol) {
      // only "http://" is supported, "https://" should use CONNECT method
      res.writeHead(400);
      res.end('Only "http:" protocol prefix is supported\n');
      return;
    }

    if (null != sOptions.localAddress) {
      // set local address for parsed args
      _localAddress = sOptions.localAddress
      _addr = ( typeof _localAddress === 'function' ) ? _localAddress() : _localAddress
      console.log('use addr %s', _addr)
      parsed.localAddress = _addr
    }

    var gotResponse = false;
    var proxyReq = http.request(parsed);
    debug.proxyRequest('%s %s HTTP/1.1 ', proxyReq.method, proxyReq.path);

    proxyReq.on('response', function (proxyRes) {
      debug.proxyResponse('HTTP/1.1 %s', proxyRes.statusCode);
      gotResponse = true;

      var headers = {};
      eachHeader(proxyRes, function (key, value) {
        debug.proxyResponse('Proxy Response Header: "%s: %s"', key, value);
        if (isHopByHop.test(key)) {
          debug.response('ignoring hop-by-hop header "%s"', key);
        } else {
          var v = headers[key];
          if (Array.isArray(v)) {
            v.push(value);
          } else if (null != v) {
            headers[key] = [ v, value ];
          } else {
            headers[key] = value;
          }
        }
      });

      debug.response('HTTP/1.1 %s', proxyRes.statusCode);
      res.writeHead(proxyRes.statusCode, headers);
      proxyRes.pipe(res);
      res.on('finish', onfinish);
    });
    proxyReq.on('error', function (err) {
      debug.proxyResponse('proxy HTTP request "error" event\n%s', err.stack || err);
      cleanup();
      if (gotResponse) {
        debug.response('already sent a response, just destroying the socket...');
        socket.destroy();
      } else if ('ENOTFOUND' == err.code) {
        debug.response('HTTP/1.1 404 Not Found');
        res.writeHead(404);
        res.end();
      } else {
        debug.response('HTTP/1.1 500 Internal Server Error');
        res.writeHead(500);
        res.end();
      }
    });

    // if the client closes the connection prematurely,
    // then close the upstream socket
    function onclose () {
      debug.request('client socket "close" event, aborting HTTP request to "%s"', req.url);
      proxyReq.abort();
      cleanup();
    }
    socket.on('close', onclose);

    function onfinish () {
      debug.response('"finish" event');
      cleanup();
    }

    function cleanup () {
      debug.response('cleanup');
      socket.removeListener('close', onclose);
      res.removeListener('finish', onfinish);
    }

    req.pipe(proxyReq);
  });
}

/**
 * HTTP CONNECT proxy requests.
 */

function onconnect (req, socket, head) {
  debug.request('%s %s HTTP/%s ', req.method, req.url, req.httpVersion);
  assert(!head || 0 == head.length, '"head" should be empty for proxy requests');

  var res;
  var target;
  var gotResponse = false;

  // define request socket event listeners
  function onclientclose (err) {
    debug.request('HTTP request %s socket "close" event', req.url);
  }
  socket.on('close', onclientclose);

  function onclientend () {
    debug.request('HTTP request %s socket "end" event', req.url);
    cleanup();
  }

  function onclienterror (err) {
    debug.request('HTTP request %s socket "error" event:\n%s', req.url, err.stack || err);
  }
  socket.on('error', onclienterror);

  // define target socket event listeners
  function ontargetclose () {
    debug.proxyResponse('proxy target %s "close" event', req.url);
    cleanup();
    socket.destroy();
  }

  function ontargetend () {
    debug.proxyResponse('proxy target %s "end" event', req.url);
    cleanup();
  }

  function ontargeterror (err) {
    debug.proxyResponse('proxy target %s "error" event:\n%s', req.url, err.stack || err);
    cleanup();
    if (gotResponse) {
      debug.response('already sent a response, just destroying the socket...');
      socket.destroy();
    } else if ('ENOTFOUND' == err.code) {
      debug.response('HTTP/1.1 404 Not Found');
      res.writeHead(404);
      res.end();
    } else {
      debug.response('HTTP/1.1 500 Internal Server Error');
      res.writeHead(500);
      res.end();
    }
  }

  function ontargetconnect () {
    debug.proxyResponse('proxy target %s "connect" event', req.url);
    debug.response('HTTP/1.1 200 Connection established');
    gotResponse = true;
    res.removeListener('finish', onfinish);

    res.writeHead(200, 'Connection established');

    // HACK: force a flush of the HTTP header
    res._send('');

    // relinquish control of the `socket` from the ServerResponse instance
    res.detachSocket(socket);

    // nullify the ServerResponse object, so that it can be cleaned
    // up before this socket proxying is completed
    res = null;

    socket.pipe(target);
    target.pipe(socket);
  }

  // cleans up event listeners for the `socket` and `target` sockets
  function cleanup () {
    debug.response('cleanup');
    socket.removeListener('close', onclientclose);
    socket.removeListener('error', onclienterror);
    socket.removeListener('end', onclientend);
    if (target) {
      target.removeListener('connect', ontargetconnect);
      target.removeListener('close', ontargetclose);
      target.removeListener('error', ontargeterror);
      target.removeListener('end', ontargetend);
    }
  }

  // create the `res` instance for this request since Node.js
  // doesn't provide us with one :(
  // XXX: this is undocumented API, so it will break some day (ノಠ益ಠ)ノ彡┻━┻
  res = new http.ServerResponse(req);
  res.shouldKeepAlive = false;
  res.chunkedEncoding = false;
  res.useChunkedEncodingByDefault = false;
  res.assignSocket(socket);

  // called for the ServerResponse's "finish" event
  // XXX: normally, node's "http" module has a "finish" event listener that would
  // take care of closing the socket once the HTTP response has completed, but
  // since we're making this ServerResponse instance manually, that event handler
  // never gets hooked up, so we must manually close the socket...
  function onfinish () {
    debug.response('response "finish" event');
    res.detachSocket(socket);
    socket.end();
  }
  res.once('finish', onfinish);

  // pause the socket during authentication so no data is lost
  socket.pause();

  authenticate(this, req, function (err, auth) {
    socket.resume();
    if (err) {
      // an error occured during login!
      res.writeHead(500);
      res.end((err.stack || err.message || err) + '\n');
      return;
    }
    if (!auth) return requestAuthorization(req, res);

    var parts = req.url.split(':');
    var host = parts[0];
    var port = +parts[1];
    var opts = { host: host, port: port };

    debug.proxyRequest('connecting to proxy target %j', opts);
    target = net.connect(opts);
    target.on('connect', ontargetconnect);
    target.on('close', ontargetclose);
    target.on('error', ontargeterror);
    target.on('end', ontargetend);
  });
}

/**
 * Checks `Proxy-Authorization` request headers. Same logic applied to CONNECT
 * requests as well as regular HTTP requests.
 *
 * @param {http.Server} server
 * @param {http.ServerRequest} req
 * @param {Function} fn callback function
 * @api private
 */

function authenticate (server, req, fn) {
  var hasAuthenticate = 'function' == typeof server.authenticate;
  if (hasAuthenticate) {
    debug.request('authenticating request "%s %s"', req.method, req.url);
    server.authenticate(req, fn);
  } else {
    // no `server.authenticate()` function, so just allow the request
    fn(null, true);
  }
}

/**
 * Sends a "407 Proxy Authentication Required" HTTP response to the `socket`.
 *
 * @api private
 */

function requestAuthorization (req, res) {
  // request Basic proxy authorization
  debug.response('requesting proxy authorization for "%s %s"', req.method, req.url);

  // TODO: make "realm" and "type" (Basic) be configurable...
  var realm = 'proxy';

  var headers = {
    'Proxy-Authenticate': 'Basic realm="' + realm + '"'
  };
  res.writeHead(407, headers);
  res.end();
}
